### PR TITLE
add descheduler framework proposal

### LINKED ISSUE #213: implementation the descheduler framework to support the event-triggered scenarios, and compatible with community frameworks

### PR DESCRIPTION
**Pull request by eahydra**
Signed-off-by: Joseph <joseph.t.lee@outlook.com>

### Ⅰ. Describe what this PR does

This proposal is based on the K8s community's [descheduler](https://github.com/kubernetes-sigs/descheduler) to design and implement the descheduler framework required by the koordinator.

### Ⅱ. Does this pull request fix one issue?

fix #213 

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+- The descheduling framework configuration can be converted into an internal representation.
+- To reduce need to specify value for every possible configuration, also defaulting serves as a recommended/opinionated settings for the plugins.
+
+#### Abstract PodEvictor interface
+
+Currently, descheduler has split `Pod Evictor` and `Evictor Filter`. Users can inject `Evictor Filter` on demand, and the plug-in calls `Evictor Filter` when selecting abnormal Pods to select Pods that meet the requirements and calls `Pod Evictor` to initiate eviction. At present, `Pod Evictor` has not been abstracted as an interface. We adopt the solution in [PoC #781](//github.com/kubernetes-sigs/descheduler/pull/781) to abstract an `Evictor interface`. And refer to [PR #885](https://github.com/kubernetes-sigs/descheduler/pull/885) to add an `EvictOptions` paramters.  We can implement custom Evictor based on [PodMigrationJob](https://github.com/koordinator-sh/koordinator/blob/main/docs/proposals/scheduling/20220701-pod-migration-job.md).
```
Comment (hormes) [inline docs/proposals/scheduling/20220701-simple-descheduler-framework.md:75]: PoC #781](https://github.com/kubernetes-sigs/descheduler/pull/781

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+We also noticed that the K8s descheduler community also found these problems and proposed corresponding solutions such as such as [#753 Descheduler framework Proposal](https://github.com/kubernetes-sigs/descheduler/issues/753) and [PoC #781](//github.com/kubernetes-sigs/descheduler/pull/781). The K8s descheduler community tries to implement a descheduler framework similar to the k8s scheduling framework. This coincides with our thinking.  
+
+On the whole, these solutions solved most of our problems, but we also noticed that the related implementations were not merged into the main branch. But we review these implementations and discussions, and we believe this is the right direction. Considering Koordinator's principle of K8s compatibility, we should also maintain compatibility with the upstream descheduler community. So when we implement the koordiantor descheduler, we will try to follow some of the designs in the #753 solution proposed by the community. When the upstream community completely transitions to the new architecture, Koordinator will follow up.
```
Comment (hormes) [inline docs/proposals/scheduling/20220701-simple-descheduler-framework.md:51]: 这里我们选用这个策略的原因，也是希望推进上游社区架构的快速演进，并不只是跟随。当然，选择在 koordinator 里先独立于社区实现的原因是我们的时间要求明确。

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md (hunk 1 above)
Comment (hormes) [reply docs/proposals/scheduling/20220701-simple-descheduler-framework.md:75]: missing https

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md (hunk 2 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-simple-descheduler-framework.md]: updated

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+## Summary
+
+This proposal is based on the K8s community's [descheduler](https://github.com/kubernetes-sigs/descheduler) to design and implement the descheduler framework required by the koordinator.
+
+## Motivation
+
+The existing [descheduler](https://github.com/kubernetes-sigs/descheduler) in the community can solve some problems, but we think that there are still many aspects of the descheduler that can be improved, for example, it only supports the mode of periodic execution, and does not support the event-triggered mode. It is not possible to extend and configure custom rescheduling strategies without invading the existing code of descheduler like kube-scheduler; it also does not support implementing custom evictor. 
+
+We also noticed that the K8s descheduler community also found these problems and proposed corresponding solutions such as such as [#753 Descheduler framework Proposal](https://github.com/kubernetes-sigs/descheduler/issues/753) and [PoC #781](https://github.com/kubernetes-sigs/descheduler/pull/781). The K8s descheduler community tries to implement a descheduler framework similar to the k8s scheduling framework. This coincides with our thinking.
```
Comment (jasonliu747) [inline docs/proposals/scheduling/20220701-simple-descheduler-framework.md:49]: duplicate `such as`

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+On the whole, these solutions solved most of our problems, but we also noticed that the related implementations were not merged into the main branch. But we review these implementations and discussions, and we believe this is the right direction. Considering that Koordiantor has clear milestones for descheduler-related features, we will implement Koordinator's own descheduler independently of the upstream community. We try to use some of the designs in the [#753 PR](https://github.com/kubernetes-sigs/descheduler/issues/753) proposed by the community and we will follow the Koordinator's compatibility principle with K8s to maintain compatibility with the upstream community descheduler when implementing. Such an independent implementation can also drive the evolution of the upstream community's work on the descheduler framework. And when the upstream community has new changes or switches to the architecture that Koordinator deems appropriate, Koordinator will follow up promptly and actively.
```
Comment (jasonliu747) [inline docs/proposals/scheduling/20220701-simple-descheduler-framework.md:51]: `Such an` -> `Such as`

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+- The descheduling framework configuration can be converted into an internal representation.
+- To reduce need to specify value for every possible configuration, also defaulting serves as a recommended/opinionated settings for the plugins.
+
+#### Abstract PodEvictor interface
+
+Currently, descheduler has split `Pod Evictor` and `Evictor Filter`. Users can inject `Evictor Filter` on demand, and the plug-in calls `Evictor Filter` when selecting abnormal Pods to select Pods that meet the requirements and calls `Pod Evictor` to initiate eviction. At present, `Pod Evictor` has not been abstracted as an interface. We adopt the solution in [PoC #781](//github.com/kubernetes-sigs/descheduler/pull/781) to abstract an `Evictor interface`. And refer to [PR #885](https://github.com/kubernetes-sigs/descheduler/pull/885) to add an `EvictOptions` paramters.  We can implement custom Evictor based on [PodMigrationJob](https://github.com/koordinator-sh/koordinator/blob/main/docs/proposals/scheduling/20220701-pod-migration-job.md).
```
Comment (jasonliu747) [inline docs/proposals/scheduling/20220701-simple-descheduler-framework.md:75]: `[PoC #781](//github.com/kubernetes-sigs/descheduler/pull/781)` -> `[PoC #781](https://github.com/kubernetes-sigs/descheduler/pull/781)`

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+type Evictor interface {
+  Evict(context.Context, *v1.Pod, options EvictOptions) bool
+}
+```
+
+#### Plug-in descheduler strategy
+
+The current descheduler has some strategies. In [PoC #781](//github.com/kubernetes-sigs/descheduler/pull/781), it is converted into `Plugin` and executed periodically. In this `periodic execution mode`, it is appropriate to abstract the policy for Pod and Node dimensions as `DeschedulePlugin` or `BalancePlugin`. The load hotspot descheduling capability that we will implement later can also implement the BalancePlugin interface.
```
Comment (jasonliu747) [inline docs/proposals/scheduling/20220701-simple-descheduler-framework.md:92]: `[PoC #781](//github.com/kubernetes-sigs/descheduler/pull/781)` -> `[PoC #781](https://github.com/kubernetes-sigs/descheduler/pull/781)`

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+On the whole, these solutions solved most of our problems, but we also noticed that the related implementations were not merged into the main branch. But we review these implementations and discussions, and we believe this is the right direction. Considering that Koordiantor has clear milestones for descheduler-related features, we will implement Koordinator's own descheduler independently of the upstream community. We try to use some of the designs in the [#753 PR](https://github.com/kubernetes-sigs/descheduler/issues/753) proposed by the community and we will follow the Koordinator's compatibility principle with K8s to maintain compatibility with the upstream community descheduler when implementing. Such an independent implementation can also drive the evolution of the upstream community's work on the descheduler framework. And when the upstream community has new changes or switches to the architecture that Koordinator deems appropriate, Koordinator will follow up promptly and actively.
+
+### Goals
+
+1. Implement Koordinator Descheduler following part of the design in #753 proposed by the community
```
Comment (jasonliu747) [inline docs/proposals/scheduling/20220701-simple-descheduler-framework.md:55]: add a link for 753?

Comment (eahydra) [reply docs/proposals/scheduling/20220701-simple-descheduler-framework.md:55]: done

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md (hunk 6 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-simple-descheduler-framework.md]: done

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md (hunk 5 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-simple-descheduler-framework.md]: done

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md (hunk 4 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-simple-descheduler-framework.md]: done

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md (hunk 3 above)
Comment (eahydra) [reply docs/proposals/scheduling/20220701-simple-descheduler-framework.md]: done

**Review by jasonliu747 — COMMENTED** [review]
/lgtm
Nice job!

File: docs/proposals/scheduling/20220701-simple-descheduler-framework.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+#### Abstract PodEvictor interface
+
+Currently, descheduler has split `Pod Evictor` and `Evictor Filter`. Users can inject `Evictor Filter` on demand, and the plug-in calls `Evictor Filter` when selecting abnormal Pods to select Pods that meet the requirements and calls `Pod Evictor` to initiate eviction. At present, `Pod Evictor` has not been abstracted as an interface. We adopt the solution in [PoC #781](https://github.com/kubernetes-sigs/descheduler/pull/781) to abstract an `Evictor interface`. And refer to [PR #885](https://github.com/kubernetes-sigs/descheduler/pull/885) to add an `EvictOptions` paramters.  We can implement custom Evictor based on [PodMigrationJob](https://github.com/koordinator-sh/koordinator/blob/main/docs/proposals/scheduling/20220701-pod-migration-job.md). 
+
+The `Evictor` interface defined as follows:
+
+```go
+type EvictOptons struct {
+  Strategy string
+	Reason   string
```
Comment (saintube) [inline docs/proposals/scheduling/20220701-simple-descheduler-framework.md:82]: wrong indent

Comment (eahydra) [reply docs/proposals/scheduling/20220701-simple-descheduler-framework.md]: fixed

**Review by jasonliu747 — COMMENTED** [review]
/lgtm

**Pull request merged**